### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by LesterOper**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 1]
```diff
@@ -1,25 +1,67 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.Objects;
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Remove unused import

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 2]
```diff
@@ -1,25 +1,67 @@
...
+        result.add(new Product(10L, "palmp", "Palm Pixi", new BigDecimal(170), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Palm/Palm%20Pixi.jpg"));
+        result.add(new Product(11L, "simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg"));
+        result.add(new Product(12L, "simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg"));
+        result.add(new Product(13L, "simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
+
+    }
     @Override
     public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+        return result.stream().filter(x-> x.getId()==id).findFirst().get();
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Compare Long values with equals(): https://stackoverflow.com/questions/20541636/comparing-boxed-long-values-127-and-128

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 3]
```diff
@@ -1,25 +1,67 @@
...
     }
-
+    
+    @Override
+    public boolean getProd(Long id){
+        return result.stream().anyMatch(x->x.getId()==id);
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: compare Long values with equals()

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 2 above)
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: What happens if there is no product with such id? Consider using .orElseThrow()

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 4]
```diff
@@ -1,25 +1,67 @@
...
+    }
     @Override
     public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+        return result.stream().filter(x-> x.getId()==id).findFirst().get();
     }
-
+    
+    @Override
+    public boolean getProd(Long id){
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Do we need this method? What is it used for?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 5]
```diff
@@ -1,25 +1,67 @@
...
-        throw new RuntimeException("Not implemented");
+        List<Product> list = result.stream().filter(x-> x.getPrice()!=null && x.getStock()>0).collect(Collectors.toCollection(ArrayList::new));
+        return list;
     }
 
     @Override
     public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+        if(this.result.stream().anyMatch(x-> x.getId()==product.getId())){
+            throw new IllegalArgumentException("Object's id isn't unic");
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Object's id isn't _unique_

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 6]
```diff
@@ -1,25 +1,67 @@
...
+        }
+        else{
+            this.result.add(product);
+        }
     }
 
     @Override
     public void delete(Long id) {
-        throw new RuntimeException("Not implemented");
+        if(this.result.stream().anyMatch(x->x.getId() == id)){
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Can we do the same without if-else block and getting by index?
.filter(...).findFirst().map(...remove product here...).orElseThrow(...)

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 7]
```diff
@@ -13,29 +13,36 @@
...
 
 public class ProductListPageServlet extends HttpServlet {
+    
+    private ProductDao product = new ArrayListProductDao();
+    
+    
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        request.setAttribute("products", getSampleProducts());
+//        List<Product> list = new ArrayList();
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java]: Remove unnecessary comments

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 8]
```diff
@@ -13,29 +13,36 @@
...
-        result.add(new Product(9L, "nokia3310", "Nokia 3310", new BigDecimal(70), usd, 100, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Nokia/Nokia%203310.jpg"));
-        result.add(new Product(10L, "palmp", "Palm Pixi", new BigDecimal(170), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Palm/Palm%20Pixi.jpg"));
-        result.add(new Product(11L, "simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg"));
-        result.add(new Product(12L, "simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg"));
-        result.add(new Product(13L, "simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
-
-        return result;
-    }
+//    private List<Product> getSampleProducts(){
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java]: Remove unnecessary comments

File: src/main/webapp/META-INF/context.xml [hunk 9]
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (SergeiStarikov) [inline src/main/webapp/META-INF/context.xml:1]: Do we need this file?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 10]
```diff
@@ -1,25 +1,67 @@
...
+    
+    @Override
+    public boolean getProd(Long id){
+        return result.stream().anyMatch(x->x.getId()==id);
+    }
+    
     @Override
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+        List<Product> list = result.stream().filter(x-> x.getPrice()!=null && x.getStock()>0).collect(Collectors.toCollection(ArrayList::new));
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Consider splitting such long lines into multiple

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 4 above)
Comment (LesterOper) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: This method is needed to test the test testDeleteProduct
It returns true if the object exists and false if not.

Comment (SergeiStarikov) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Please consider renaming the method then.
Two methods with similar names: getProd() and getProduct() - looks ambiguous

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java [hunk 11]
```diff
@@ -2,20 +2,41 @@
...
+    
     @Before
     public void setup() {
         productDao = new ArrayListProductDao();
     }
 
     @Test
     public void testFindProductsNoResults() {
-        assertTrue(productDao.findProducts().isEmpty());
+        assertTrue(!productDao.findProducts().isEmpty());
```
Comment (SergeiStarikov) [inline src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java]: Can be simplified to assertFalse?
assertFalse(productDao.findProducts().isEmpty())

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java [hunk 12]
```diff
@@ -2,20 +2,41 @@
...
+    public void testSaveProduct(){
+        Product prod = new Product(14L, "nokia111", "Nokia 355", new BigDecimal(50), Currency.getInstance("USD"), 10, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Samsung/Samsung%20Galaxy%20S.jpg" );
+        productDao.save(prod);
+        assertTrue(productDao.getProduct(prod.getId())!=null);
+    }
+    
+    @Test
+    public void testDeleteProduct(){
+        productDao.delete(10L);
+        assertTrue(productDao.getProd(10L)==false);
```
Comment (SergeiStarikov) [inline src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java]: assertEquals instead?

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java [hunk 13]
```diff
@@ -2,20 +2,41 @@
...
 
     @Test
     public void testFindProductsNoResults() {
-        assertTrue(productDao.findProducts().isEmpty());
+        assertTrue(!productDao.findProducts().isEmpty());
+    }
+    
+    @Test
+    public void testGetProduct(){
+        assertTrue(productDao.getProduct(4L)!=null);
```
Comment (SergeiStarikov) [inline src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java]: assertNotNull instead?

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java [hunk 14]
```diff
@@ -2,20 +2,41 @@
...
+    }
+    
+    @Test
+    public void testSaveProduct(){
+        Product prod = new Product(14L, "nokia111", "Nokia 355", new BigDecimal(50), Currency.getInstance("USD"), 10, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Samsung/Samsung%20Galaxy%20S.jpg" );
+        productDao.save(prod);
+        assertTrue(productDao.getProduct(prod.getId())!=null);
+    }
+
```
Comment (SergeiStarikov) [inline src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:37]: Consider adding more tests on edge cases: for example, test to add two products with the same id.

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 4 above)
Comment (LesterOper) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Ok
But if I use assertEquals, then this method will not be necessary.

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 2 above)
Comment (LesterOper) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: What do plus and minus mean?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 4 above)
Comment (SergeiStarikov) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: That's the point. 

There are two very similar methods:

public Product getProduct(Long id) {
        return result.stream().filter(x-> x.getId()==id).findFirst().get();
    }

public boolean getProd(Long id){
        return result.stream().anyMatch(x->x.getId()==id);
    }

They both use similar code to get the product by id. The first returns the product, the second returns the boolean value. Can we get rid of of of these methods? Or call one from another?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 15]
```diff
@@ -4,7 +4,7 @@
 import java.util.ArrayList;
 import java.util.Currency;
 import java.util.List;
-import java.util.Objects;
+
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Remove empty line

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 16]
```diff
@@ -4,7 +4,7 @@
 import java.util.ArrayList;
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java:4]: Please use descriptive commit messages in future like:

"Implemented ProductDao"
"Fixed code accorindg to comments"

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 17]
```diff
@@ -30,38 +30,40 @@ public ArrayListProductDao(){
...
-    
-    @Override
-    public boolean getProd(Long id){
-        return result.stream().anyMatch(x->x.getId()==id);
+        result.stream().filter(x-> x.getId().equals(id)).findFirst().
+                orElseThrow(NullPointerException::new);
+        
+        return result.stream().
+                filter(x-> x.getId().equals(id)).
+                findFirst().get();
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Remove code duplication:

return  ...findFirst().orElseThrow() ?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 18]
```diff
@@ -30,38 +30,40 @@ public ArrayListProductDao(){
...
     @Override
     public void save(Product product) {
-        if(this.result.stream().anyMatch(x-> x.getId()==product.getId())){
-            throw new IllegalArgumentException("Object's id isn't unic");
+        if(this.result.stream().anyMatch(x-> x.getId().equals(product.getId()))){
+            
+            this.result.stream().
+                filter(x-> x.getId().equals(product.getId())).
+                    findFirst().map(x-> this.result.set(this.result.indexOf(x), product));
         }
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: The wrong indentation makes code difficult to read. Please compare:

this.result.stream().
                filter(x-> x.getId().equals(product.getId())).
                    findFirst().map(x-> this.result.set(this.result.indexOf(x), product));

AND

this.result.stream()
        .filter(x-> x.getId().equals(product.getId()))
        .findFirst()
        .map(x-> this.result.set(this.result.indexOf(x), product));

Please have a look: https://marcin-chwedczuk.github.io/java-streams-best-practices

Please check all such places in code.

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 19]
```diff
@@ -30,38 +30,40 @@ public ArrayListProductDao(){
...
+                collect(Collectors.toCollection(ArrayList::new));
         return list;
     }
 
     @Override
     public void save(Product product) {
-        if(this.result.stream().anyMatch(x-> x.getId()==product.getId())){
-            throw new IllegalArgumentException("Object's id isn't unic");
+        if(this.result.stream().anyMatch(x-> x.getId().equals(product.getId()))){
+
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: unnecessary empty line? please check all such places

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 20]
```diff
@@ -7,42 +7,30 @@
...
+    }
     
     
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-//        List<Product> list = new ArrayList();
-//        list.add(product.getProduct(5L));
-          //product.delete(12L);
-        request.setAttribute("products",product.findProducts() );
+        synchronized(this){
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java]: This is a bit tricky.

Please check spec: https://download.oracle.com/otn-pub/jcp/servlet-3.0-fr-eval-oth-JSpec/servlet-3_0-final-spec.pdf?AuthParam=1572334946_7f4786ec37dbfbce193ba15d6cc4e3cd - 2.3.3.1 "Multithreading Issues"

And look through the docs:
https://books.google.by/books?id=Wf5ihBC7gucC&pg=PA185&lpg=PA185&dq=synchronize+doget+dopost&source=bl&ots=hVdR-3iUll&sig=ACfU3U0n0yOw3jKh8C2ZK5ml_n_WjYor3Q&hl=en&sa=X&ved=2ahUKEwiph6GY_MDlAhUDwsQBHULGCHwQ6AEwDnoECAkQAQ#v=onepage&q&f=false

https://tomcat.apache.org/tomcat-5.5-doc/servletapi/javax/servlet/http/HttpServlet.html

https://books.google.by/books?id=YC0geTPQt24C&pg=PA266&lpg=PA266&dq=synchronize+doget+dopost&source=bl&ots=wcQeQF4Hjb&sig=ACfU3U2j2SZSGf1lkrsq_ory2Ke_lxMj8g&hl=en&sa=X&ved=2ahUKEwiph6GY_MDlAhUDwsQBHULGCHwQ6AEwEXoECAcQAQ#v=onepage&q&f=false

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 21]
```diff
@@ -7,42 +7,30 @@
...
+            try{
+                Thread.sleep(5000);
+            }
+            catch(InterruptedException e){
+                System.err.println("Interrupted");
+            }
+        }
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java]: Really no need for this.

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 10 above)
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: The variable list is redundant. Just return the expression.

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 22]
```diff
@@ -1,25 +1,65 @@
...
+    public synchronized void save(Product product) {
+        if(this.result.stream().anyMatch(x-> x.getId().equals(product.getId()))){
+            this.result.stream().
+                filter(x-> x.getId().equals(product.getId())).
+                    findFirst().
+                    map(x-> this.result.set(this.result.indexOf(x), product));
+        }
+        else{
+            this.result.add(product);
+        }
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java:55]: Can we simplify this to:

```
if (products.contains(product)) {
  ...throw exception...
}
if (product!=null) {
  ...add product...
}

```
OR with streams:

```
if (products.stream()
    .anyMatch(...compapre ID here...) {
    ...throw exception...
} else {
    ...add product...
}
```
